### PR TITLE
Improve performance of variable resolver

### LINKED ISSUE #456: Optimising `find_var` / `find_arg` performance
I've got a large JS program which `qjs` takes ~1.5 seconds just to parse/load. If I profile the evaluation using `gprof`, then it shows that over a third of the time is spent in [`find_var`](https://github.com/quickjs-ng/quickjs/blob/07fa1cbc4a8938085a716051c0f14a3453a885fb/quickjs.c#L20252-L20260) alone:

```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls   s/call   s/call  name    
 36.65      1.84     1.84   275818     0.00     0.00  find_var
 12.75      2.48     0.64   212052     0.00     0.00  JS_CallInternal
 11.55      3.06     0.58   100362     0.00     0.00  get_var_ref
...
```

Given that every time`find_var` is called it iterates through all variables in the scope until finding the target variable, it looks like this is causing a lot of repeated iteration over the same variables when `fd->var_count` is large (15951 in my case):

```c
static int find_var(JSContext *ctx, JSFunctionDef *fd, JSAtom name)
{
    int i;
    for(i = fd->var_count; i-- > 0;) {
        if (fd->vars[i].var_name == name && fd->vars[i].scope_level == 0)
            return i;
    }
    return find_arg(ctx, fd, name);
}
```

Could this be replaced with a lookup/hashtable or similar in the `JSFunctionDef` object? Or is there a simpler alternative?

### PR DESCRIPTION
**Pull request by bnoordhuis**
Switch to a hash table when the number of variables grows beyond a threshold.

Speeds up the test case from the linked issue by about 70%.

Fixes: https://github.com/quickjs-ng/quickjs/issues/456

### CODE REVIEW

**Review by saghul — APPROVED** [review]

**Comment by bnoordhuis** [comment]
I'm going to tweak it a little more before I land it, stay tuned.

File: quickjs.c [hunk 1]
```diff
@@ -20514,6 +20515,86 @@ static __exception int emit_push_const(JSParseState *s, JSValue val,
...
 
+// perl hash; variation of k&r hash with a different magic multiplier
+// and a final shuffle to improve distribution of the low-order bits
+static uint32_t hash_bytes(uint32_t h, const void *b, size_t n)
+{
+    const char *p;
+
+    for (p = b; p < (char *)b + n; p++)
+        h = 33*h + *p;
+    h += h >> 5;
```
Comment (bnoordhuis) [inline quickjs.c:20526]: Switching the hash function resulted in a whopping 15% fewer hash collisions. Pretty big deal.

Catastrophic collisions are still going to be an issue once the table grows to a few million elements but I hope and assume that's never going to happen in the real world. Even that monster from the linked issue has < 20,000 elements.

**Review by saghul — APPROVED** [review]
Nice! Random question: we seem to have a few htables around now, do you see an opportunity to DRY some out? Not in this PR of course.

**Comment by richarddd** [comment]
> Nice! Random question: we seem to have a few htables around now, do you see an opportunity to DRY some out? Not in this PR of course.

This would be great. Maybe even pulling in something external? https://github.com/JacksonAllan/Verstable :D

**Comment by bnoordhuis** [comment]
> we seem to have a few htables around now, do you see an opportunity to DRY some out?

Dunno. They're all just different enough from each other that it's not a slam dunk DRY job.

I did notice that, compared to perl hash, `shape_hash` and `get_index_hash` have like 5x the number of hash collisions on the (limited) inputs I tested it with.

Swapping the hash function didn't make a measurable difference in my quick testing though, likely because the shape and IC hash tables just don't get that big.

**Pull request merged**